### ISSUE TITLE
Buffer and group parallel task output.

### ISSUE THREAD

**Issue by brad-jones** [go-task/task#104]
At the moment when many commands are run as dependencies, they run in parallel and their output is all mixed up and hard to read. Would be great to add similar functionality to GNU parallel.

> With --keep-order --line-buffer will output lines from the first job while it is running, then lines from the second job while that is running. It will buffer full lines, but jobs will not mix.  
https://www.gnu.org/software/parallel/man.html

For example:
```
$ parallel --line-buffer --keep-order ::: "ping -4 2 8.8.8.8" "ping -c 2 8.8.4.4"
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
64 bytes from 8.8.8.8: icmp_seq=1 ttl=53 time=17.3 ms
64 bytes from 8.8.8.8: icmp_seq=2 ttl=53 time=17.2 ms
64 bytes from 8.8.8.8: icmp_seq=3 ttl=53 time=14.9 ms
64 bytes from 8.8.8.8: icmp_seq=4 ttl=53 time=15.4 ms

--- 8.8.8.8 ping statistics ---
4 packets transmitted, 4 received, 0% packet loss, time 3004ms
rtt min/avg/max/mdev = 14.969/16.238/17.307/1.061 ms
PING 8.8.4.4 (8.8.4.4) 56(84) bytes of data.
64 bytes from 8.8.4.4: icmp_seq=1 ttl=54 time=16.8 ms
64 bytes from 8.8.4.4: icmp_seq=2 ttl=54 time=15.9 ms

--- 8.8.4.4 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1001ms
rtt min/avg/max/mdev = 15.999/16.443/16.888/0.462 ms
```

And/Or the ability to tag each line, something like this:

```
$ parallel --line-buffer --rpl '{color} $_="\033[30;3".(++$::color%8)."m"' --tagstring {color}[{}] ::: "ping -c 4 8.8.8.8" "ping -c 2 8.8.4.4"
[ping -c 4 8.8.8.8]	PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
[ping -c 4 8.8.8.8]	64 bytes from 8.8.8.8: icmp_seq=1 ttl=53 time=16.0 ms
[ping -c 2 8.8.4.4]	PING 8.8.4.4 (8.8.4.4) 56(84) bytes of data.
[ping -c 2 8.8.4.4]	64 bytes from 8.8.4.4: icmp_seq=1 ttl=54 time=16.8 ms
[ping -c 4 8.8.8.8]	64 bytes from 8.8.8.8: icmp_seq=2 ttl=53 time=16.7 ms
[ping -c 2 8.8.4.4]	64 bytes from 8.8.4.4: icmp_seq=2 ttl=54 time=16.5 ms
[ping -c 2 8.8.4.4]	
[ping -c 2 8.8.4.4]	--- 8.8.4.4 ping statistics ---
[ping -c 2 8.8.4.4]	2 packets transmitted, 2 received, 0% packet loss, time 1001ms
[ping -c 2 8.8.4.4]	rtt min/avg/max/mdev = 16.580/16.728/16.876/0.148 ms
[ping -c 4 8.8.8.8]	64 bytes from 8.8.8.8: icmp_seq=3 ttl=53 time=18.0 ms
[ping -c 4 8.8.8.8]	64 bytes from 8.8.8.8: icmp_seq=4 ttl=53 time=18.0 ms
[ping -c 4 8.8.8.8]	
[ping -c 4 8.8.8.8]	--- 8.8.8.8 ping statistics ---
[ping -c 4 8.8.8.8]	4 packets transmitted, 4 received, 0% packet loss, time 3003ms
[ping -c 4 8.8.8.8]	rtt min/avg/max/mdev = 16.092/17.261/18.097/0.876 ms
```

**Comment by andreynering** [comment]
@brad-jones Thanks for the suggestion, I think that makes sense, and it's not hard to implement.

Of these two examples, I prefer the first one.

What would you prefer for the UX for this: a cli flag or an YAML option in the task?

**Comment by brad-jones** [comment]
Yes first example is my preference also.

Personally I can't really think of a case where you would want some dependencies to run with the current "unbuffered" output and some with the new "buffered" output. Therefore a CLI flag seems logical.

I am obviously biased but it would be great if this new "buffered" output was something you had to opt out of, instead of opt in.

**Comment by andreynering** [comment]
> Personally I can't really think of a case where you would want some dependencies to run with the current "unbuffered" output and some with the new "buffered" output. Therefore a CLI flag seems logical.

Hmm... maybe a global config on the Taskfile, so.

> I am obviously biased but it would be great if this new "buffered" output was something you had to opt out of, instead of opt in.

There are good reasons for the current behavior. I think it's good to have live feedback for a slow compilation, for example. Or someone running a daemon (a command that don't ends) like a web application server (that will print request logs to stdout/stderr), etc.

**Comment by brad-jones** [comment]
Yes your second point is a good one, I am yet to try this but have set this sort of thing up in the past with other technologies.

I assume if one was to have a task with only `deps` and those were long running tasks, ie: http server, web socket server and file watcher then `task` would not die until told to or the deps exited for whatever reason.

In this sort of situation, tagging the lines, as per the second example would be helpful.

**Comment by smyrman** [comment]
Adding it to the Taskfile should involve releasing a new point version. E.g.:
```
version: "2.1"
parallelOutput: "group|prefix-interleaved|interleaved"
```
@andreynering, what is the code strategy for such an addition?

An override flag that could potentially work also for 2.0 Taskfiles (or even 1.0 Taskfiles):
```
--parllelOutput= "group|prefixinterleaved|interleaved"
```
?

commands like `docker-compose logs` use the line tag strategy.

**Comment by smyrman** [comment]
> @andreynering, what is the code strategy for such an addition?

**Comment by andreynering** [comment]
@smyrman 

> what is the code strategy for such an addition?

Buffered output would be easier, maybe we should implement that first. Line tags requires a bit more thought on the implementation.

Maybe we improve how we handle versions in the future, but for now can be a simple if and returning an error.

I think maybe we should allow this option both in a task and a global in the Taskfile. We can also have a flag, though, if you guys think it's useful.

**Comment by brad-jones** [comment]
> I think maybe we should allow this option both in a task and a global in the Taskfile. We can also have a flag, though, if you guys think it's useful.

Sounds good to me. If your going to have a global yaml option, don't think you then need a cli flag as well. Too many ways to configure something creates ambiguity.

```yaml
parallelOutput: "group|prefix-interleaved|interleaved"
```
I like this, if "prefix-interleaved" is too hard right now, that can always come later.

The prefixes can just be the task names, no need to include the entire command like my gnu parallel example.
